### PR TITLE
add mysql_full_version and suffix return variable

### LINKED ISSUE #114: mysql_user: get full mysql version string as a variable
##### SUMMARY
Right now if you use `mysql_info` to get the version of the database, you get back a dict with three items for the major, minor and release version. However it's hard to use this variable in a version-comparison (e.g. `when:  mysql_version is version('5.6', '>=')`), because you first have to construct a new variable from the dict.

Therefore I propose to add a new variable (e.g. `full_mysql_version`) that contains the full version string.

One open question for me: would this be a new filter or should this new variable be returned in the `version`-filter?

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
mysql_info

##### ADDITIONAL INFORMATION

Sample code:

```
    def __get_global_variables(self):
        """Get global variables (instance settings)."""
        res = self.__exec_sql('SHOW GLOBAL VARIABLES')

        if res:
            for var in res:
                self.info['settings'][var['Variable_name']] = self.__convert(var['Value'])

            ver = self.info['settings']['version'].split('.')
            release = ver[2].split('-')[0]

            self.info['version'] = dict(
                major=int(ver[0]),
                minor=int(ver[1]),
                release=int(release),
            )

+            self.info['full_version']  = self.info['settings']['version']
```

### PR DESCRIPTION
**Pull request by rndmh3ro**
##### SUMMARY
This PR adds two new variables: the `suffix` variable that is sometimes added to the version-string and the `full_version`-variable that contains the whole version string (to make it easier to compare versions).

Fixes #114 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
mysql_info

### CODE REVIEW

**Review by Andersson007 — COMMENTED** [review]
@rndmh3ro thanks for the PR
Please  also add a changelog fragment https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment

File: plugins/modules/mysql_info.py [hunk 1]
```diff
@@ -138,6 +139,16 @@
       returned: if not excluded by filter
       type: int
       sample: 60
+    suffix:
+      description: Server suffix, e.g. MySQL, MariaDB, other or none.
```
Comment (Andersson007) [inline plugins/modules/mysql_info.py]:
```suggestion
      description: Server suffix, for example, MySQL, MariaDB, other or none.
```
It's from Ansible best practices guide

**Comment by Andersson007** [comment]
fyi: don't worry about `codecov/project` is red.
Waiting for the changelog fragment

**Comment by rndmh3ro** [comment]
I added the fragment!

File: changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- mysql_info - add mysql_full_version and suffix return variable (https://github.com/ansible-collections/community.mysql/issues/114)
```
Comment (Andersson007) [inline changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml]:
```suggestion
- mysql_info - add ``full_version`` and ``suffix`` return variable (https://github.com/ansible-collections/community.mysql/issues/114).
```
to highlight a little bit. we should also put a dot at the end

Comment (Andersson007) [reply changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml]: and maybe `variableS`? but i'm not an expert in English:)

Comment (Andersson007) [reply changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml]: or maybe even `values`. It's just a suggestion

Comment (rndmh3ro) [reply changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml]: You're right, I changed it, thanks!

**Review by Andersson007 — APPROVED** [review]
LGTM
@Jorge-Rodriguez if the PR looks good to you too, feel free to merge

**Comment by Jorge-Rodriguez** [comment]
@rndmh3ro could you add an explicit integration test for the suffix output?
Currently we're testing only against MySQL do it should always be empty, and it'll help cover the use case when we finally at MariaDB integration testing.

Thanks.

File: plugins/modules/mysql_info.py [hunk 3]
```diff
@@ -121,7 +122,7 @@
   description: Database server version.
   returned: if not excluded by filter
   type: dict
-  sample: { "version": { "major": 5, "minor": 5, "release": 60 } }
+  sample: { "version": { "major": 5, "minor": 5, "release": 60, "suffix": "MariaDB", "full_version": "5.5.60-MariaDB" } }
```
Comment (Andersson007) [inline plugins/modules/mysql_info.py]:
```suggestion
  sample: { "version": { "major": 5, "minor": 5, "release": 60, "suffix": "MariaDB", "full": "5.5.60-MariaDB" } }
```
@rndmh3ro maybe just `full`? IMO its invocation `version.full` in playbooks would look better than `version.full_version`.
what do you think?

Comment (Andersson007) [reply plugins/modules/mysql_info.py]: @rndmh3ro BTW if you like to contribute more, you could add `full` (but w/o `suffix` IMO) to https://github.com/ansible-collections/community.postgresql/blob/main/plugins/modules/postgresql_info.py and https://github.com/ansible-collections/community.postgresql/blob/main/plugins/modules/postgresql_ping.py 
What do you think? (Or let me know if you're busy, i could do it)

File: plugins/modules/mysql_info.py [hunk 4]
```diff
@@ -138,6 +139,16 @@
       returned: if not excluded by filter
       type: int
       sample: 60
+    suffix:
+      description: Server suffix, for example MySQL, MariaDB, other or none.
+      returned: if not excluded by filter
+      type: str
+      sample: "MariaDB"
+    full_version:
```
Comment (Andersson007) [inline plugins/modules/mysql_info.py]:
```suggestion
    full:
```

File: plugins/modules/mysql_info.py [hunk 5]
```diff
@@ -353,13 +364,30 @@ def __get_global_variables(self):
             for var in res:
                 self.info['settings'][var['Variable_name']] = self.__convert(var['Value'])
 
-            ver = self.info['settings']['version'].split('.')
-            release = ver[2].split('-')[0]
+            # version = ["5", "5," "60-MariaDB]
+            version = self.info['settings']['version'].split('.')
+
+            # full_version = "5.5.60-MariaDB"
+            full_version = self.info['settings']['version']
```
Comment (Andersson007) [inline plugins/modules/mysql_info.py]:
```suggestion
            full = self.info['settings']['version']
```

File: plugins/modules/mysql_info.py [hunk 6]
```diff
@@ -353,13 +364,30 @@ def __get_global_variables(self):
...
             self.info['version'] = dict(
-                major=int(ver[0]),
-                minor=int(ver[1]),
+                # major = "5"
+                major=int(version[0]),
+                # minor = "5"
+                minor=int(version[1]),
                 release=int(release),
+                suffix=str(suffix),
+                full_version=str(full_version),
```
Comment (Andersson007) [inline plugins/modules/mysql_info.py]:
```suggestion
                full=str(full),
```

File: tests/integration/targets/test_mysql_info/tasks/main.yml [hunk 7]
```diff
@@ -48,7 +48,7 @@
     - assert:
         that:
         - result.changed == false
-        - result.version != {}
+        - "mysql_version in result.version.full_version"
```
Comment (Andersson007) [inline tests/integration/targets/test_mysql_info/tasks/main.yml]:
```suggestion
        - "mysql_version in result.version.full"
```

File: changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- mysql_info - add `full_version` and `suffix` return values (https://github.com/ansible-collections/community.mysql/issues/114).
```
Comment (Andersson007) [inline changelogs/fragments/115-add_mysql_full_version_suffix_return_var.yml]:
```suggestion
- mysql_info - add `version.full` and `version.suffix` return values (https://github.com/ansible-collections/community.mysql/issues/114).
```

**Comment by rndmh3ro** [comment]
@Andersson007, thanks I'll add the changes locally.

>  Currently we're testing only against MySQL do it should always be empty, and it'll help cover the use case when we finally at MariaDB integration testing.

@Jorge-Rodriguez sadly it's not always empty. Looking at the CI-logs:

MySQL 8:
```
2021-03-12T08:50:26.7204419Z TASK [test_mysql_info : Collect info only about version and databases] *********
2021-03-12T08:50:27.1526986Z [0;32mok: [testhost] => {"changed": false, "databases": {"information_schema": {"size": 0}, "mysql": {"size": 2654208}, "performance_schema": {"size": 0}, "sys": {"size": 16384}}, "version": {"full_version": "8.0.22", "major": 8, "minor": 0, "release": 22, "suffix": ""}}[0m


2021-03-12T08:51:25.3928912Z TASK [test_mysql_replication : find out the database version] ******************
2021-03-12T08:51:25.9604764Z [0;32mok: [testhost] => {"changed": false, "version": {"full_version": "8.0.22", "major": 8, "minor": 0, "release": 22, "suffix": ""}}[0m
```

MySQL 5.7:
```
2021-03-12T08:49:45.3867627Z TASK [test_mysql_info : Collect info only about version and databases] *********
2021-03-12T08:49:45.8869227Z [0;32mok: [testhost] => {"changed": false, "databases": {"information_schema": {"size": 163840}, "mysql": {"size": 2644595}, "performance_schema": {"size": 0}, "sys": {"size": 16384}}, "version": {"full_version": "5.7.31-log", "major": 5, "minor": 7, "release": 31, "suffix": "log"}}[0m

2021-03-12T08:51:00.7500011Z TASK [test_mysql_replication : find out the database version] ******************
2021-03-12T08:51:01.4887762Z [0;32mok: [testhost] => {"changed": false, "version": {"full_version": "5.7.31-log", "major": 5, "minor": 7, "release": 31, "suffix": "log"}}[0m
```

How should we proceed here?

**Comment by Jorge-Rodriguez** [comment]
@rndmh3ro maybe a unit test would be better suited for the task

**Comment by rndmh3ro** [comment]
> maybe a unit test would be better suited for the task

I'm not familiar with writing unit tests. My Python-knowledge in this regard is quite limited. How does one do that? :)

**Comment by Jorge-Rodriguez** [comment]
> > maybe a unit test would be better suited for the task
> 
> I'm not familiar with writing unit tests. My Python-knowledge in this regard is quite limited. How does one do that? :)

There are a couple of simple examples in the tests/units folder. If you still don't feel comfortable, just let me know and I can do it.

**Comment by Andersson007** [comment]
@Jorge-Rodriguez it's a very good spot about unit tests.
If @rndmh3ro is not familiar with it, i wouldn't overwhelm him with it (but it's good to study how pytest work in general, just see tests/units/* content).
But if you like to figure out the topic, we could wait.
If not, i think i could implement the unit tests later (or anyone else if they wish), after this is merged if there's no concerns that the changes will break the module.
@Jorge-Rodriguez if you put possible input and output that we expect here, it would be really great.

What do you folks think?

(@rndmh3ro feel free to fix what i suggested for a while.)

**Comment by Andersson007** [comment]
@Jorge-Rodriguez ah, didn't see your comment, cool, thanks!

**Comment by rndmh3ro** [comment]
I'd be glad if any of you two can write the unit test (and after that I'll try to understand it :)).

**Comment by Andersson007** [comment]
@rndmh3ro sounds sensible

**Comment by Jorge-Rodriguez** [comment]
@rndmh3ro That unit test took a lot more magick than I expected. I'll be happy to go through it with you if you want.
@Andersson007 I included the changes you requested.

**Comment by rndmh3ro** [comment]
>  I'll be happy to go through it with you if you want.

That'd be great. There are some parts I don't understand.
Either here or on IRC? On here it's probably better for discoverability so others could learn from it, too. However maybe this PR is not the right place for this.

**Comment by Jorge-Rodriguez** [comment]
@rndmh3ro I'll be hanging out as Tiriel on the Ansible channel on freenode

**Review by Andersson007 — APPROVED** [review]
LGTM
@Jorge-Rodriguez thanks for the unit tests and adding the suggested changes!
Feel free to merge

**Comment by Jorge-Rodriguez** [comment]
Better yet would be to review the unit test in a repo discussion. @Andersson007 can we enable those?

**Comment by Andersson007** [comment]
@Jorge-Rodriguez , enabled, please check

**Comment by Andersson007** [comment]
@rndmh3ro there's more complex case in `community.postresql`. I'd add it myself there. I'll put a reference to this PR with a comment that the original idea is yours.

**Comment by Jorge-Rodriguez** [comment]
> @Jorge-Rodriguez , enabled, please check

It works. Thanks

**Comment by rndmh3ro** [comment]
https://github.com/ansible-collections/community.mysql/discussions/122

**Pull request merged**